merkle-patricia-tree/src/Data/Tree/MerklePatricia.hs

```diff
@@ -381,14 +418,110 @@ emptyRoot = hashNode "" []
+--
+-- The first part of the proof being considered is (100110, []) which
+-- represents the leaf and the suffix of the reference (or the leaf's prefix).
+-- There's no children and no value, because the value is provided by the caller
+-- for verification. From there, we can easily calculate the resulting leaf hash
+-- by re-hashing the value and the corresponding part of the suffix and then go
+-- recursively through each node levels above, in reverse order.
+--
+-- The next one is (ø, [(1,38e90)]) which indeed correspond to the _other_branch
+-- once we have consumed '100110' from the end of the reference. The path to the
+-- element is on the branch `0`, thus we need the branch `1` to continue the
+-- proof.
+--
+-- This is also where it gets interesting. Calculating the root hash for
+-- the sub-MPT _with_ the element is trivial. We can re-construct the hash for the
+-- branch on our path, and the proof contains the hash of the other branch.
+-- Calculating the hash of the sub-MPT _without_ the element is more complex
+-- because, the structure of the tree changes. When the element is present, we
+-- have a Node with no prefix, and two branches on '0' and '1'. When the element
+-- is not present, we have a leaf, with '1000110' as prefix.
+--
+-- What is perhaps taken for granted at the moment (and quite convenient you'll
+-- admit!) is that, the hash of the Leaf with prefix 1000110 and the hash of the
+-- node from the branch '1' with a single leaf with prefix 000110 are actually
+-- the same!
+-- This is done _by construction_ to make it possible to reconstruct hashes even
+-- if the structure of the tree changes! For more details, see the doc-string on
+-- the 'hashPrefix' function, but in short: a prefix is hashed in multiple rounds,
+-- one hash per digit, all nested. Such that it's easy to add or peel a layer
+-- from the oignon-hash. To get a sense of this, imagine for a moment that this
+-- wasn't the case, and that the hash of the leaf would be different than the
+-- hash of a node with leaf... how would you reconstruct the hash of the MPT
+-- with the target element removed? When constructed "naively", we would typically
+-- construct the node hash as:
+--
+--        *--- Node's prefix
+--        v
+--   hash(ø|hash(000110|hash(value)))
+--                 ^
+--                 *--- Leaf's prefix
+--
+-- But now, to reconstruct the tree _without_ the element, we would need to
+-- construct (still naively) the hash as:
+--
+--           *--- Leaf's prefix
+--           v
+--   hash(1000110|hash(value))
+--
+-- We don't have hash(value) in the proof, and we don't really have a way to get
+-- from the former to the latter. So, the trick in the construction of the tree
+-- and all the hashes is to (a) include the branch digit in the hash and (b) hash
+-- every digit separately, in an oignon pattern such that, the proof has
+-- actually the following shape:
+--
+--   hash(1|hash(0|hash(0|hash(0|hash(1|hash(1|hash(0|hash(value))))))))
```

Comment:
This section is in fact a bit verbose here. It would suffice to say that by hashing intelligently (see `hashPrefix`) we are able to even handle re-structured tries.